File: LeetCode/Medium/1143. Longest Common Subsequence.java
class Solution
{

    public int longestCommonSubsequence( String text1, String text2 )
    {
        if ( text1.length() < text2.length() )
        {
            return longestCommonSubsequence( text2, text1 );
        }
        
        int n = text2.length() + 1;
        int[] p = new int[n];

        for ( int i = 0; i < text1.length(); i++ )
        {
            int[] q = new int[n];

            for( int j = 1; j < n; j++ )
            {
                if ( text1.charAt( i ) == text2.charAt( j - 1 ) )
                {
                    q[j] = p[ j - 1 ] + 1;
                }
                else
                {
                    q[j] = Math.max( p[j], q[ j - 1 ] );
                }
                
            }
            
            p = q;
        }
        
        return p[ n - 1 ];
    }

}

File: LeetCode/Medium/1143. Longest Common Subsequence.cs
public class Solution
{

    public int LongestCommonSubsequence( string text1, string text2 )
    {
        if ( text1.Length < text2.Length )
        {
            return LongestCommonSubsequence( text2, text1 );
        }
        
        int n = text2.Length + 1;
        int[] p = new int[n];

        for ( int i = 0; i < text1.Length; i++ )
        {
            int[] q = new int[n];

            for( int j = 1; j < n; j++ )
            {
                if ( text1[i] == text2[ j - 1 ] )
                {
                    q[j] = p[ j - 1 ] + 1;
                }
                else
                {
                    q[j] = Math.Max( p[j], q[ j - 1 ] );
                }
                
            }
            
            p = q;
        }
        
        return p[ n - 1 ];
    }

}
